vms/platformvm/validators/manager.go

```diff
@@ -164,100 +164,138 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
 			return nil, err
 		}
 	}
+	currentSubnetValidatorList := currentSubnetValidators.List()
+	subnetSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentSubnetValidatorList))
+	for _, vdr := range currentSubnetValidatorList {
+		subnetSet[vdr.NodeID] = &validators.GetValidatorOutput{
+			NodeID: vdr.NodeID,
+			// PublicKey will be picked from primary validators
+			Weight: vdr.Weight,
+		}
+	}
+
 	currentPrimaryNetworkValidators, ok := m.cfg.Validators.Get(constants.PrimaryNetworkID)
 	if !ok {
 		// This should never happen
 		return nil, ErrMissingValidator
 	}
-
-	currentSubnetValidatorList := currentSubnetValidators.List()
-	vdrSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentSubnetValidatorList))
-	for _, vdr := range currentSubnetValidatorList {
-		primaryVdr, ok := currentPrimaryNetworkValidators.Get(vdr.NodeID)
-		if !ok {
-			// This should never happen
-			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, vdr.NodeID)
-		}
-		vdrSet[vdr.NodeID] = &validators.GetValidatorOutput{
+	currentPrimaryValidatorList := currentPrimaryNetworkValidators.List()
+	primarySet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentPrimaryValidatorList))
+	for _, vdr := range currentPrimaryValidatorList {
+		primarySet[vdr.NodeID] = &validators.GetValidatorOutput{
 			NodeID:    vdr.NodeID,
-			PublicKey: primaryVdr.PublicKey,
+			PublicKey: vdr.PublicKey,
 			Weight:    vdr.Weight,
 		}
+
+		// fill PK from primary network
+		if _, found := subnetSet[vdr.NodeID]; found {
+			subnetSet[vdr.NodeID].PublicKey = vdr.PublicKey
+		}
 	}
```

Comment:
Instead of populating the keys as nil, then populating them with the primary network's keys i think we can populate them all in one go if we do this:
```
	for _, vdr := range currentPrimaryValidatorList {
                 validator := &validators.GetValidatorOutput{
				NodeID:    vdr.NodeID,
				PublicKey: vdr.PublicKey,
				Weight:    vdr.Weight,
		}

		if currentSubnetValidators.Contains(vdr.NodeID) {
			subnetValidatorSet[vdr.NodeID] = validator
		}

		primarySet[vdr.NodeID] = validator
	}
```

Reply:
Looks like we can get rid of the subnetSet -> list overhead if we do this as well

Reply:
@joshua-kim each validator's weight here is the sum of validator's own weight and its delegators. So a subnet validator's weight is diffrent than its primary validator's weight. For this reasons I think the change does not work

Reply:
Ah I see

Reply:
Reverted this change